### PR TITLE
[GSoC] Common Python Simulation module

### PR DESCRIPTION
**Pull request by harshkhandeparkar**
Related to #211

#### Changes
- [x] Added a common Python simulation module for use across all generators.
- [x] Used the simulation module to simplify and improve the Temperature Sensor generator simulations.
- [x] Added documentation (both in Python and in `docs/` that is hosted on ReadTheDocs) for the module.
- [x] Regression Tests

#### Features of the Simulation Module
- Easy to sweep through multiple parameters.
- Generates the output in a well-defined directory structure.
- Supports [Mako](https://makotemplates.org) templating syntax for inserting parameters in SPICE templates (just like in Verilog templates).
- Runs concurrent simulations.
- Reports the simulation status.

Simulations for the DC-DC generator will be based on this.

#### Screenshots
image

### CODE REVIEW

**Comment by harshkhandeparkar** [comment]
@saicharan0112 any update on this?

**Review by saicharan0112 — APPROVED** [review]
LGTM

File: .github/workflows/test_code.yml
Comment (msaligane) [inline .github/workflows/test_code.yml:1]: can this file be renamed in a better way?

File: docs/source/common-python-api.rst
Comment (msaligane) [inline docs/source/common-python-api.rst:1]: We should probably rename this as well.

File: openfasoc/generators/common/simulation/simulation_run.py
Comment (msaligane) [inline openfasoc/generators/common/simulation/simulation_run.py:1]: can we add some descriptions here (other files too)

File: openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp
Comment (msaligane) [inline openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp:1]: Are we using Xyce?

File: openfasoc/generators/temp-sense-gen/tools/readparamgen.py [hunk 1]
```diff
@@ -5,7 +5,6 @@
 import os
 import re
 import sys
-import time
```
Comment (msaligane) [inline openfasoc/generators/temp-sense-gen/tools/readparamgen.py:8]: Why is this removed?

File: tests/common_api/test-simulation/expected/3/sim_3.sp
Comment (msaligane) [inline tests/common_api/test-simulation/expected/3/sim_3.sp:1]: These are regression tests for the common-api? the naming is kind of weird.

**Review by msaligane — CHANGES_REQUESTED** [review]

File: openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp
Comment (harshkhandeparkar) [reply openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp:1]: There is a option to use it (a CLI argument) but I am not sure if it is actually used anywhere or if it is working.

File: openfasoc/generators/temp-sense-gen/tools/readparamgen.py (hunk 1 above)
Comment (harshkhandeparkar) [reply openfasoc/generators/temp-sense-gen/tools/readparamgen.py:8]: The library was not used so I removed the import statement.

File: tests/common_api/test-simulation/expected/3/sim_3.sp
Comment (harshkhandeparkar) [reply tests/common_api/test-simulation/expected/3/sim_3.sp:1]: These are the expected SPICE files generated from `template.sp`. There are 8 possible combinations of parameters, and hence 8 different configurations are generated.

Comment (harshkhandeparkar) [reply tests/common_api/test-simulation/expected/3/sim_3.sp:1]: The regression test checks if the correct configurations are generated by comparing the file generated during the testing to the expected file.

File: openfasoc/generators/common/simulation/simulation_run.py
Comment (harshkhandeparkar) [reply openfasoc/generators/common/simulation/simulation_run.py:1]: Done.

File: .github/workflows/test_code.yml
Comment (saicharan0112) [reply .github/workflows/test_code.yml:1]: Yes that shouldn't be a problem

File: openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp
Comment (saicharan0112) [reply openfasoc/generators/temp-sense-gen/simulations/templates/tempsenseInst_xyce.sp:1]: We are installing Xyce as part of the installation script. I remember having problems with the simulations using Xyce but I don't know if there is any progress on that side in general inside OpenFASoC

File: tests/common_api/test-simulation/expected/3/sim_3.sp
Comment (saicharan0112) [reply tests/common_api/test-simulation/expected/3/sim_3.sp:1]: @msaligane I don't think these are actually a problem.

File: openfasoc/generators/temp-sense-gen/test.json
Comment (saicharan0112) [inline openfasoc/generators/temp-sense-gen/test.json:1]: Can we rename this file please? I think `config.json` would be the right name.

**Review by saicharan0112 — COMMENTED** [review]
@harshkhandeparkar I think the refactoring the simulation part of the generators can be made compatible with the CI. My only concern is the independence of the functions that were previously in separate files such as `result.py` and `result_error.py`. Are these refactored functions standalone and can be initialized directly in a different python file or does it expect any declarations that are only known inside the newly developed `simulation.py` ?

File: openfasoc/generators/temp-sense-gen/test.json
Comment (harshkhandeparkar) [reply openfasoc/generators/temp-sense-gen/test.json:1]: I think renaming this could break many things since test.json is used in almost all generators and in many places to pull the user parameters. It would be better if this is done in a separate PR.

Comment (saicharan0112) [reply openfasoc/generators/temp-sense-gen/test.json:1]: I am referring to the file `test.json` under the temp-sense-gen so I don't think other generators actually read this file. Probably making changes to the part of the code where the reading actually happens will do the job. 

But maybe it can be done in a different PR.

Comment (harshkhandeparkar) [reply openfasoc/generators/temp-sense-gen/test.json:1]: > I don't think other generators actually read this file.

Sorry, I meant to say that the naming is used in other generators.

> Probably making changes to the part of the code where the reading actually happens will do the job. But maybe it can be done in a different PR.

Yes, that would be better since it's read in many places. Changing it will make this PR too long.

**Comment by harshkhandeparkar** [comment]
> @harshkhandeparkar I think the refactoring the simulation part of the generators can be made compatible with the CI. My only concern is the independence of the functions that were previously in separate files such as `result.py` and `result_error.py`. Are these refactored functions standalone and can be initialized directly in a different python file or does it expect any declarations that are only known inside the newly developed `simulation.py` ?

As far as I remember, the functionality is intact and modular. The functions were moved (and might be renamed) to `simulation_error.py`, which is now a Python module that can be imported but doesn't run anything. Could you tell which functions from the previous files were required? If they are not available anymore, I'll add them back.

**Comment by msaligane** [comment]
@harshkhandeparkar can you pull the changes from the main repo.

**Comment by harshkhandeparkar** [comment]
> @harshkhandeparkar can you pull the changes from the main repo.

Done.

**Review by msaligane — APPROVED** [review]

**Comment by saicharan0112** [comment]
@harshkhandeparkar looks like the simulation module changes are not working from the beginning? (another fault of CI is detected thanks to this). not sure how the python tests didn't see this as well. Can you try running the temp-sense-gen again at your side please? Just run `make temp-sense-gen` and you should be able to see the error while importing the `calculate_error...` function from `simulation_result.py` inside the main `temp-sense-gen.py` file

**Comment by harshkhandeparkar** [comment]
> @harshkhandeparkar looks like the simulation module changes are not working from the beginning? (another fault of CI is detected thanks to this). not sure how the python tests didn't see this as well. Can you try running the temp-sense-gen again at your side please? Just run `make temp-sense-gen` and you should be able to see the error while importing the `calculate_error...` function from `simulation_result.py` inside the main `temp-sense-gen.py` file

I just tested it and it works on my system locally but it does not work in the `openfasoc` Docker container. I believe this is because the container uses Python version `3.7.x`. Is it possible to bump the version?

Also I think I made a big mistake in #233 which caused the CI to not run at all.

**Comment by saicharan0112** [comment]
> > @harshkhandeparkar looks like the simulation module changes are not working from the beginning? (another fault of CI is detected thanks to this). not sure how the python tests didn't see this as well. Can you try running the temp-sense-gen again at your side please? Just run `make temp-sense-gen` and you should be able to see the error while importing the `calculate_error...` function from `simulation_result.py` inside the main `temp-sense-gen.py` file
> 
> I just tested it and it works on my system locally but it does not work in the `openfasoc` Docker container. I believe this is because the container uses Python version `3.7.x`. Is it possible to bump the version?
> 
> Also I think I made a big mistake in #233 which caused the CI to not run at all.

Yes but I handled it in #240 . Regarding the python version, I am not sure. The entire environment is kind of pinned to 3.7. I think there is a way to use 3.8 (something that I am looking into #237) but 3.10 is something I am not sure about.

**Comment by saicharan0112** [comment]
I think I am able to find a way to install all tools in python 3.10. I still need to test the generators with latest openroad (not sure how big that will blow) but I will update the docker image to give you the python version that you need for your code.

**Comment by harshkhandeparkar** [comment]
> I think I am able to find a way to install all tools in python 3.10. I still need to test the generators with latest openroad (not sure how big that will blow) but I will update the docker image to give you the python version that you need for your code.

I can try to make it compatible with `3.7`.

**Comment by saicharan0112** [comment]
> > I think I am able to find a way to install all tools in python 3.10. I still need to test the generators with latest openroad (not sure how big that will blow) but I will update the docker image to give you the python version that you need for your code.
> 
> I can try to make it compatible with `3.7`.

no 3.7 is quite old and miniconda not supporting 3.7 is a sign for developers not use it anymore. Also since now all dependencies are working in 3.10 env, there should be no problem in handling such stuff.
